builtin/credential/aws-iam/path_login.go

```diff
@@ -0,0 +1,376 @@
+		Fields: map[string]*framework.FieldSchema{
+			"role": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `Name of the role against which the login is being attempted.
+If 'role' is not specified, then the login endpoint looks for a role
+bearing the name of the IAM principal that is trying to login in.
+If a matching role is not found, the login fails.`,
+			},
+			"method": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `HTTP method to use for the AWS request. This must match what
+has been signed in the presigned request. Currently, POST is the only supported value`,
+			},
+			"url": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `Full URL against which to make the AWS request. If using a POST
+request with the action specified in the body, this should just be
+"/".`,
+			},
+			"body": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `Base64-encoded request body. This must match the request
+body included in the signature.`,
+			},
+			"headers": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `Base64-encoded JSON representation of the request headers.
+This must at a minimum include the headers over which AWS has included a
+signature.`,
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathLoginUpdate,
+		},
+
+		HelpSynopsis:    pathLoginSyn,
+		HelpDescription: pathLoginDesc,
+	}
+}
+
+func (b *backend) pathLoginUpdate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	// BEGIN boring data parsing
+	method := data.Get("method").(string)
+	if method == "" {
+		return logical.ErrorResponse("missing method"), nil
+	}
+
+	// In the future, might consider supporting GET
+	if method != "POST" {
+		return logical.ErrorResponse("Invalid method; currently only 'POST' is supported"), nil
+	}
+
+	rawUrl := data.Get("url").(string)
+	if rawUrl == "" {
+		return logical.ErrorResponse("missing url"), nil
+	}
+	parsedUrl, err := url.Parse(rawUrl)
+	if err != nil {
+		return logical.ErrorResponse("error parsing url"), nil
+	}
+
+	// TODO: There are two potentially valid cases we're not yet supporting that would
+	// necessitate this check being changed. First, if we support GET requests.
+	// Second if we support presigned POST requests
+	bodyB64 := data.Get("body").(string)
+	if bodyB64 == "" {
+		return logical.ErrorResponse("missing body"), nil
+	}
+	bodyRaw, err := base64.StdEncoding.DecodeString(bodyB64)
+	if err != nil {
+		return logical.ErrorResponse("body is invalid base64"), nil
+	}
+	body := string(bodyRaw)
+
+	headersB64 := data.Get("headers").(string)
+	if headersB64 == "" {
+		return logical.ErrorResponse("missing headers"), nil
+	}
+	headersJson, err := base64.StdEncoding.DecodeString(headersB64)
+	if err != nil {
+		return logical.ErrorResponse("headers is invalid base64"), nil
+	}
+	var headers map[string]string
```

Comment:
The more I think about this, the more I think you're probably right. Vault is in go, the code should use go idioms as much as possible. Give me a few days to rewrite this using the go-standard format and also see what the python client looks like and I'll either have an updated PR or some actual code showing why this would be uglier :)